File: PythonSandbox/src/leetcode/lc133_clone_graph.py
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        
        # iterative BFS
        q = [node]
        copyMap = {node.val: Node(node.val)}

        while q:
            currNode = q.pop(0)
            currNodeCopy = copyMap[currNode.val]

            for _,neighbor in enumerate(currNode.neighbors):
                if neighbor.val not in copyMap.keys():
                    copyMap[neighbor.val] = Node(neighbor.val)
                    q.append(neighbor)
                
                neighborCopy = copyMap[neighbor.val]
                currNodeCopy.neighbors.append(neighborCopy)

        return copyMap[node.val]

File: JavaSandbox/src/leetcode/LC133_CloneGraph.java
package leetcode;

import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;

// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

class LC133_CloneGraph {
    public Node cloneGraph(Node node) {
        if(node == null) {
            return null;
        }

        List<Node> q = new ArrayList<>();
        q.add(node);
        HashMap<Integer, Node> copyMap = new HashMap<>();
        copyMap.put(node.val, new Node(node.val));

        while(q.size() > 0) {
            Node currNode = q.get(0);
            q.remove(0);
            //System.out.println("currNode removed: " + currNode);
            Node currNodeCopy = copyMap.get(currNode.val);

            for(Node neighbor: currNode.neighbors) {
                if(!copyMap.keySet().contains(neighbor.val)) {
                    copyMap.put(neighbor.val, new Node(neighbor.val));
                    q.add(neighbor);
                }

                Node neighborCopy = copyMap.get(neighbor.val);
                currNodeCopy.neighbors.add(neighborCopy);
            }
        }

        return copyMap.get(node.val);
    }
}
